nominal/core/nominal_write_stream.py

```diff
@@ -0,0 +1,104 @@
+    ) -> None:
+        """Exit the stream and close out any used system resources."""
+
+    @abc.abstractmethod
+    def enqueue(
+        self,
+        channel_name: str,
+        timestamp: str | datetime | IntegralNanosecondsUTC,
+        value: float | str,
+        tags: dict[str, str] | None = None,
+    ) -> None:
+        """Write a single value to the stream
+
+        Args:
+            channel_name: Name of the channel to upload data for.
+            timestamp: Absolute timestamp of the data being uploaded.
+            value: Value to write to the specified channel.
+            tags: Key-value tags associated with the data being uploaded.
+                NOTE: This *must* include all `required_tags` used when creating a `Connection` to Nominal.
+        """
+
+    def enqueue_batch(
+        self,
+        channel_name: str,
+        timestamps: Sequence[str | datetime | IntegralNanosecondsUTC],
+        values: Sequence[float | str],
+        tags: dict[str, str] | None = None,
+    ) -> None:
+        """Add a sequence of messages to the queue to upload to Nominal.
+
+        Messages are added one-by-one (with timestamp normalization) and flushed
+        based on the batch conditions.
+
+        Args:
+            channel_name: Name of the channel to upload data for.
+            timestamps: Absolute timestamps of the data being uploaded.
+            values: Values to write to the specified channel.
+            tags: Key-value tags associated with the data being uploaded.
+                NOTE: This *must* include all `required_tags` used when creating a `Connection` to Nominal.
+        """
+        if len(timestamps) != len(values):
+            raise ValueError(
+                f"Expected equal numbers of timestamps and values! "
+                f"Received {len(timestamps)} timestamp(s) vs. {len(values)} value(s)."
+            )
+
+        for timestamp, value in zip(timestamps, values):
+            self.enqueue(channel_name, timestamp, value, tags)
+
+    def enqueue_from_dict(
+        self,
+        timestamp: str | datetime | IntegralNanosecondsUTC,
+        channel_values: dict[str, float | str],
+    ) -> None:
+        """Write multiple channel values at a given timestamp using a flattened dictionary.
+
+        Each key in the dictionary is treated as a channel name and the corresponding value
+        is enqueued with the given timestamp.
+
+        Args:
+            timestamp: The shared timestamp to use for all items to enqueue.
+            channel_values: A dictionary mapping channel names to their respective values.
+        """
+        for channel, value in channel_values.items():
+            self.enqueue(channel, timestamp, value)
+
+    @abc.abstractmethod
+    def flush(self, wait: bool = False, timeout: float | None = None) -> None:
+        """Flush current batch of records to nominal.
+
```

Comment:
not related to this PR but we should specify this is NOT REQUIRED to be called manually